msgraph-developer-proxy/ConsoleLogger.cs

```diff
@@ -7,30 +7,315 @@ namespace Microsoft.Graph.DeveloperProxy;
 
 public class ConsoleLogger : ILogger {
     private readonly ConsoleColor _color;
+    private readonly LabelMode _labelMode;
+    private readonly LogLevel _logLevel;
+    private readonly string _boxTopLeft = "\u250c ";
+    private readonly string _boxLeft = "\u2502 ";
+    private readonly string _boxBottomLeft = "\u2514 ";
+    // used to align single-line messages
+    private readonly string _boxSpacing = "  ";
 
-    public ConsoleLogger() {
+    private static readonly object lockObject = new object();
+
+    public ConsoleLogger(ProxyConfiguration configuration) {
         _color = Console.ForegroundColor;
+        _labelMode = configuration.LabelMode;
+        _logLevel = configuration.LogLevel;
     }
 
-    public void Log(string message) {
+    public void LogInfo(string message) {
+        if (_logLevel > LogLevel.Info) {
+            return;
+        }
+
         Console.WriteLine(message);
     }
 
     public void LogWarn(string message) {
+        if (_logLevel > LogLevel.Warn) {
+            return;
+        }
+
         Console.ForegroundColor = ConsoleColor.Yellow;
-        Console.Error.WriteLine($"\tWARNING: {message}");
+        Console.Error.WriteLine($"  WARNING: {message}");
         Console.ForegroundColor = _color;
     }
 
     public void LogError(string message) {
+        if (_logLevel > LogLevel.Error) {
+            return;
+        }
+
         Console.ForegroundColor = ConsoleColor.Red;
         Console.Error.WriteLine(message);
         Console.ForegroundColor = _color;
     }
 
     public void LogDebug(string message) {
+        if (_logLevel > LogLevel.Debug) {
+            return;
+        }
+
         Console.ForegroundColor = ConsoleColor.Gray;
         Console.Error.WriteLine(message);
         Console.ForegroundColor = _color;
     }
+
+    public void LogRequest(string[] message, MessageType messageType, LoggingContext? context = null) {
+        var messageLines = new List<string>(message);
+
+        // add request context information to the message for messages
+        // that are not intercepted requests and have a context
+        if (messageType != MessageType.InterceptedRequest &&
+            context is not null) {
+            messageLines.Add($"{context.Session.HttpClient.Request.Method} {context.Session.HttpClient.Request.Url}");
+        }
+
+        lock (lockObject) {
```

Comment:
We need lock to avoid broken rendering:

<img width="1278" alt="image" src="https://user-images.githubusercontent.com/11164679/213157607-0e4421f4-4a96-4a50-97a7-72dda5609837.png">